src/main/scala/loamstream/model/jobs/JobDirs.scala

```diff
@@ -0,0 +1,110 @@
+package loamstream.model.jobs
+
+import scala.collection.Seq
+import loamstream.util.{ Paths => LPaths }
+import loamstream.util.{ Files => LFiles }
+import loamstream.util.Sequence
+import java.nio.file.Path
+import java.nio.file.Paths.{ get => path }
+import java.nio.file.Files
+import scala.util.Try
+
+
+object JobDirs {
+
+  def allocate(jobs: Iterable[LJob], branchingFactor: Int, nextId: () => String = () => DirNode.nextId()): DirNode = {
+    require(branchingFactor > 1)
+    require(jobs.nonEmpty)
+    
+    if(jobs.size <= branchingFactor) {
+      DirNode.Interior(id = nextId(), children = jobs.map(DirNode.Leaf(_)))
+    } else {
+      val numChunks = branchingFactor
+      
+      import Implicits.SeqOps      
+      
+      DirNode.Interior(
+          id = nextId(), 
+          children = jobs.toSeq.splitInto(numChunks).map(chunk => allocate(chunk, branchingFactor, nextId)))
+    }
+  }
+  
+  sealed trait DirNode {
+    def toSimplePathName: String = LPaths.mungePathRelatedChars {
+      this match {
+        case l: DirNode.Leaf => l.job.name
+        case i: DirNode.Interior => i.id
+      }
+    }
+    
+    def pathsByJob(jobDataDir: Path): Map[LJob, Path] = {
+      import LPaths.Implicits._
+
+      def doPathsByJob(root: Path)(n: DirNode): Iterable[(LJob, Path)] = n match {
+        case l @ DirNode.Leaf(job) => Seq(job -> (root / l.toSimplePathName).normalize)
+        case i @ DirNode.Interior(id, children) => {
+          val nodeRoot = root / i.toSimplePathName
+          
+          children.flatMap(doPathsByJob(nodeRoot))
+        }
+      }
+      
+      doPathsByJob(jobDataDir)(this).toMap
+    }
+    
+    def makeDirsUnder(root: Path): Boolean = {
+      import LPaths.Implicits._
+      import LFiles.createDirsIfNecessary
+      
+      val pathsToCreate = pathsByJob(root).values.iterator
+      
+      val attempts = pathsToCreate.map(p => Try(createDirsIfNecessary(p)))
+      
+      attempts.forall(_.isSuccess)
+    }
+  }
+  
+  object DirNode {
+    private[this] val lock = new AnyRef
+    
+    private[this] val ids: Iterator[String] = {
+      val alphabet = "abcdefghijklmnopqrstuvwxyz".toUpperCase
```

Comment:
Trivial, but I goof up with things like this and miss characters all the time. Suggestion:

```scala
val alphabet = ('A' to 'Z').mkString
```

Reply:
Cool, that's much better!  The actual letters used don't mean anything, but it would be lame to skip one due to a typo.